clang/lib/AST/ExprConstant.cpp

```diff
@@ -7095,6 +7096,40 @@ class APValueToBufferConverter {
     return true;
   }
 
+  bool visitVector(const APValue &Val, QualType Ty, CharUnits Offset) {
+    const auto *VT = Ty->castAs<VectorType>();
+    unsigned VectorLength = Val.getVectorLength();
+
+    if (VT->isExtVectorBoolType()) {
+      // Special handling for OpenCL bool vectors: we need to pack the vector
+      // of 1-bit unsigned integers into a single integer with the corresponding
+      // bits set, then write out the resulting integer.
+
+      CharUnits VecWidthBits = Info.Ctx.getTypeSizeInChars(VT) * 8;
+
+      APSInt Bits(VecWidthBits.getQuantity());
+      for (unsigned I = 0; I != VectorLength; ++I) {
+        const APValue &SubObj = Val.getVectorElt(I);
+        assert(SubObj.isInt() && "Bool vector element isn't an int?");
+        Bits.setBitVal(I, !SubObj.getInt().isZero());
```

Comment:
Is this endianness-dependent? (I have no idea, but it seems worth checking.)

Reply:
It shouldn't be, once the value's been loaded the byte order should be irrelevant. The byte order conversion (if any) appears to be handled automagically by `llvm::StoreIntToMemory` (and its `load` counterpart).

Reply:
From what I've found through some experimentation, it looks like:

- Vector element 0 is always stored at a lower address than vector element 8.
- The actual vector loads and stores operate like integer loads and stores, and in particular, if the vector is narrower than the storage, you get padding beyond the high-order bits.

Which I think means that for big-endian systems we want to put vector element `I` in integer bit `VectorLength - I - 1`, so that when we reverse and right-align the bits, we end up with the first vector element in the earlier memory address.

That seems to match what's described in https://llvm.org/docs/LangRef.html#vector-type

Reply:
Hm, it concerns me that the documentation says: "When <N*M> isn’t evenly divisible by the byte size the exact memory layout is unspecified (just like it is for an integral type of the same size). This is because different targets could put the padding at different positions when the type size is smaller than the type’s store size."

Can we reliably constant-evaluate a `bit_cast` from a vector of a number of bits that's not a multiple of the byte size?

Reply:
I've come to the same conclusion as you that I don't think this is possible with non-byte-sized vectors. Since this is only an issue for OpenCL bool vectors, should I just have it fail with an unsupported exception when the number of bool elements isn't a multiple of the char size?

Reply:
I think that'd be fine. It doesn't seem like those vectors are well-supported elsewhere in Clang and LLVM (for example, I hit a miscompile bug #67060 while trying to figure out how they are stored in memory), so I think it's fine if they're also not well-supported by constant evaluation, at least for now.

Reply:
Okay, the latest revision will now fail to evaluate vectors which aren't a multiple of the byte size.